godot-core/src/builtin/variant/mod.rs

```diff
@@ -288,25 +301,70 @@ impl Variant {
+        unsafe { std::slice::from_raw_parts(variant_ptr_array, length) }
     }
 
-    /// Converts to variant mut pointer; can be a null pointer.
-    pub(crate) fn ptr_from_sys_mut(variant_ptr: sys::GDExtensionVariantPtr) -> *mut Variant {
-        variant_ptr as *mut Variant
+    /// Convert an array of `Variant` sys pointers to a slice of mutable `Variant` references.
+    ///
+    /// # Safety
+    ///
+    /// Either `variant_array` is null, or it must be safe to call [`std::slice::from_raw_parts`] with
+    /// `variant_array` cast to `*const Variant` and `length`.
+    pub(crate) unsafe fn borrow_var_slice<'a>(
+        variant_array: sys::GDExtensionConstVariantPtr,
+        length: usize,
+    ) -> &'a [Variant] {
+        sys::static_assert_eq_size_align!(Variant, sys::types::OpaqueVariant);
+
+        // Godot may pass null to signal "no arguments" (e.g. in custom callables).
+        if variant_array.is_null() {
+            debug_assert_eq!(
+                length, 0,
+                "Variant::unbounded_refs_from_sys(): pointer is null but length is not 0"
+            );
+            return &[];
+        }
+
+        // raw pointers and references have the same memory layout.
+        // See https://doc.rust-lang.org/reference/type-layout.html#pointers-and-references-layout.
+        let variant_array = variant_array.cast::<Variant>();
+
+        // SAFETY: `variant_array` isn't null so it is safe to call `from_raw_parts` on the pointer cast to `*const Variant`.
+        unsafe { std::slice::from_raw_parts(variant_array, length) }
     }
 
-    /// Move `self` into a system pointer. This transfers ownership and thus does not call the destructor.
+    /// Convert an array of `Variant` sys pointers to a slice of mutable `Variant` references.
     ///
     /// # Safety
-    /// `dst` must be a pointer to a [`Variant`] which is suitable for ffi with Godot.
-    pub(crate) unsafe fn move_var_ptr(self, dst: sys::GDExtensionVariantPtr) {
-        self.move_return_ptr(dst as *mut _, sys::PtrcallType::Standard);
+    ///
+    /// Either `variant_array` is null, or it must be safe to call [`std::slice::from_raw_parts_mut`] with
+    /// `variant_array` cast to `*mut Variant` and `length`.
+    pub(crate) unsafe fn borrow_var_slice_mut<'a>(
+        variant_array: sys::GDExtensionVariantPtr,
+        length: usize,
+    ) -> &'a mut [Variant] {
+        sys::static_assert_eq_size_align!(Variant, sys::types::OpaqueVariant);
+
+        // Godot may pass null to signal "no arguments" (e.g. in custom callables).
+        if variant_array.is_null() {
+            debug_assert_eq!(
+                length, 0,
+                "Variant::unbounded_refs_from_sys(): pointer is null but length is not 0"
+            );
+            return &mut [];
+        }
+
+        // raw pointers and references have the same memory layout.
```

Comment:
See above.

Btw, if this comment appears multiple times, we should maybe mention it once before all functions.